### ISSUE TITLE
Dockerfile doesn't actually build.

### ISSUE THREAD

**Issue by preston** [cqframework/cqf-ruler#202]
Would you accept a PR for making the Dockerfile actually build the project? It doesn't really do anything except `COPY` in the build, which doesn't work for automated CI systems.

**Comment by preston** [comment]
FYI this would be for HL DaVinci having hosted versions at Logica.

**Comment by JPercival** [comment]
Hi Preston,

We actually used to do it that way but found it was not suited for our purposes. There were two main issues:

1. We deliver the cqf-ruler in a number of different ways. As a docker image, as a java jar, and as a war. Having the full build within the docker container meant either building the same artifact several times or copying the built artifacts out of the docker container. Building it outside of the final container and copying it in allows us to ensure the same binary is used for all use cases, and eliminates the risk of the build environment getting out of sync for those scenarios.

2. Docker does not allow build-time mounting of volumes. This means that build is done completely from scratch each time, with all the dependencies being downloaded from Maven on each build. This makes build times needlessly long since most of the dependencies could be cached (and in fact, on Travis we do use a cache to speed up the build). This also reduces build stability since there are occasional network hiccups or outages with Maven central. Having the dependencies cached makes the build succeed whether or not Maven is up.

When researching alternatives we found that this was a common issue for Java applications. A common solution is to break the build pipeline into two separate docker containers. One with the build tools already installed and could have the cache and source code mounted at runtime which performs the actual build and a second that's the runtime container.

The setup is relatively easy to do in docker based CI scenarios. Here's one guide on how to do it: [Java Docker Pipeline](https://codefresh.io/docker-tutorial/java_docker_pipeline/)

The short version:
1.Mount the source into a maven image and run the build
```bash
docker run --rm --name cqf-ruler-build -u $(id -u ${USER}):$(id -g ${USER}) -v "$(pwd)":/usr/src/cqf-ruler -w /usr/src/cqf-ruler maven:3.6-jdk-11 mvn clean package -DskipTests=true
```
2. Run the cqf-ruler Docker image build

Hope this helps!

**Comment by preston** [comment]
Docker's default behavior of `docker build … ` is to cache individual layers unless some prior input has changed. It only builds from scratch if the `--no-cache` option is explicitly used, a prior change was noted, or something has cleared the local cache. This is the same for Java applications as all others. If you're seeing "from scratch" behavior every time, it's probably because application source isn't properly copied in _after_ pom.xml dependencies are installed, or an misconfigured .dockerignore. Those issues are almost always the problem for all languages and is usually a pretty easy fix. The solution is essentially the same for all languages and frameworks.

Having a full build process that is completely consistent and reproducible, regardless of local environment, is largely the point! Anyone wanting to simply build/run the service using the current method has to actually set up and maintain a local development environment as opposed to using whatever is declared in the Dockerfile, at which point the Dockerfile itself is largely pointless.

Also, the major cloud platforms are optimized to operate this way. It shouldn't require language/framework-specific CI to do builds. Building externally removes some of the best benefits of containerization. Multi-phase builds across bases images has also been solved for a while.

I'm happy to submit a PR … otherwise every end user with a deployment (like us) will have to either maintain a special external compilation pipeline or maintain their own fork with a more de facto Dockerfile, and deal with manually merging out to keep up. Neither should really be necessary though.

Thanks for all your work on this .. let me know what you think.

**Comment by JPercival** [comment]
Hi Preston

Thanks for you response. Let me address some of the points you've raised.

> Docker's default behavior of docker build … is to cache individual layers unless some prior input has changed.

This behavior is dependent upon the CI platform. If the layers are not cached between builds this is not the behavior that will be observed. In particular, Travis CI does not do this without some additional steps since it spins up a fresh VM on every build. Here's a thread about it:

https://github.com/travis-ci/travis-ci/issues/5358

> Having a full build process that is completely consistent and reproducible, regardless of local environment, is largely the point! 

It's my belief that if a single image build counts a consistent and reproducible so would a multi-image build like I proposed in my last message.

> Also, the major cloud platforms are optimized to operate this way. It shouldn't require language/framework-specific CI to do builds

Travis CI, the build tool we are using for the cqf-ruler, does not behave this way as I linked above. In my previous message, I linked a second CI product that always don't not behave that way.

>  Multi-phase builds across bases images has also been solved for a while.

The previous approach was a multi-stage build. It's still in the commit history if it's something you really need. Here's a link: [Old Dockerfile](https://github.com/DBCG/cqf-ruler/blob/e68a12336e59565c7f50de5d6b34c576d8ab032b/Dockerfile)

Docker is not our only delivery pipeline, so I don't believe that focusing on it exclusively makes sense for our needs.

>  otherwise every end user with a deployment (like us) will have to either maintain a special external compilation pipeline

The Docker image itself is available at contentgroup/cqf-ruler. There's no need to do a build at if you just want to run the cqf-ruler. If you are maintaining a custom fork, manually merging is already something you are presumably handling.

I'm not opposed to having a single Dockerfile in the repo. But it's not something that will address our particular needs and our particular delivery pipelines, so I don't foresee it as something we'd maintain.

**Comment by preston** [comment]
Requiring installation of develpment dependencies on individuals' machines most certainly is not a consistent way of creating reproducable builds. I understand using Travis for CI. This is about CD, not CI. It sounds like the disconnect is that cqf-ruler is only concerned with the CI mechanisms. No one is advocating for only using containerized releases.

For those doing CD -- or more concerned about the deployment side of developing against cqf-ruler -- we aren't able to rely on the bundled Dockerfile since it doesn't actually build as one would expect.

**Comment by JPercival** [comment]
Hi Preston,

> Requiring installation of develpment dependencies on individuals' machines

There must be some misunderstanding. That is simply _not the case_. There is a docker image with the build tools installed (specifically maven). One must only have docker installed, the same as if it were all in one Dockerfile. Admittedly, the docs are quite out of date.

> This is about CD, not CI.

We have multiple clients using the cqf-ruler in a CD environment. Here's sample deployspec.yml for AWS: 

```yaml
version: 0.2

phases:
  install:
    runtime-versions:
      docker: 18
  post_build:
    commands:
      - docker run --rm -u $(id -u ${USER}):$(id -g ${USER}) -v "$(pwd)":/usr/src/cqf-ruler -w /usr/src/cqf-ruler maven:3.6-jdk-11 mvn clean package -DskipTests=true
      - docker build -t  /path/to/custom/registry/cqf-ruler:$IMAGE_TAG .
      - docker push /path/to/custom/registry/cqf-ruler:$IMAGE_TAG
   .... (whatever other steps are needed for deployments)...
```
(Obviously whether or not that's a proper deployspec could be argued -- my point is that it can be done using only Docker. A buildspec file would look similar)

I can also come up with Kubernetes and docker-compose examples, if that's helpful.

Since the cqf-ruler is not a container-native (or even container-first - we have clients running it on legacy windows servers in Tomcat) application and we must support both use cases that's left us with essentially three options:

1. Maintain separate build and deployment pipelines for docker vs war deployments
    * This is where we started, and it bit us because there's the risk of the build environments getting out of sync and the corresponding binaries getting out of sync.
2. Attempt keeping the build pipeline all in docker, and then copy the binaries out of the image for war-based deployments
   *  The process for copying artifacts out of an image is harder than copying them in, since there's first-class support for copying things into an image with the COPY directive of a Docker buildfile.
   * Maven has no plugin capable of deploying from a pre-built binary, as far as I'm aware. If you attempt a mvn deploy with a binary you've copied from somewhere else it'll attempt to rebuild.
3. Share the build part of the pipeline, split the deploy part of the pipeline.
   * This is where the ended up. A maven docker image that will build and deploy the war, and then copying the war into a docker image which can be used for deployment.

If there's a better way to share build tools and binaries across legacy and docker deployments I'm all ears. I would very much like to simplify things. As it stands, you can use a maven build image in every docker-based CD environment of which I'm aware and it works fine. There's no need to install build tools on every machine.
